### PR TITLE
Introduce `RefasterMethodParameterOrder` check

### PR DESCRIPTION
**Pull request by Stephan202**
:exclamation: This PR is on top of #962 :exclamation:

Another step towards canonical Refaster rule definitions.

Suggested commit message:
```
Introduce `RefasterMethodParameterOrder` check (#775)

While there, simplify some `EqualityRules` Refaster rules, as a suggested
parameter order change highlighted that they could be collapsed.
```

### CODE REVIEW

**Comment by Stephan202** [comment]
Rebased and added a commit to resolve a new violation.

**Comment by Stephan202** [comment]
Rebased and resolved conflict.

**Review by mohamedsamehsalah — APPROVED** [review]
Clean 🧹

**Review by rickie — APPROVED** [review]
Took some time to wrap my head around the algorithm, but it looks really nice 🚀 !

One question 👀.

File: error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/RefasterMethodParameterOrderTest.java
Comment (rickie) [reply error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/RefasterMethodParameterOrderTest.java:43]: Part of the reason that I put this comment here is that we could also choose to use `expectNoDiagnostics` here instead (?) of `doTest` to make the intention of the test clearer.

**Comment by rickie** [comment]
Hmm, interesting: 

```
Error:  Failures: 
Error:    RefasterMethodParameterOrderTest.replacement:119 compilation failed unexpectedly: [/17892554101468984197/ARecipes.java:1: error: ';' expected
package unnamed package;
               ^]
```

**Comment by Stephan202** [comment]
I can reproduce that locally, also when upgrading to the currently-latest `template-writing` (version 1.4.1).

Also, when I run the tests in IDEA (in an attempt to debug the problem :upside_down_face:), I get the same issue as mentioned in https://github.com/PicnicSupermarket/error-prone-support/pull/925#issuecomment-1873305005; it seems that issue breaks ~all tests in IDEA, not just `StringRulesRecipesTest` :grimacing:.

This'll require a closer look (can't promise when; busy agenda). CC @timtebeek as an FYI.

**Comment by Stephan202** [comment]
I traced the IDEA issue to [IDEA-342187](https://youtrack.jetbrains.com/issue/IDEA-342187), for which I filed #958. This should unblock debugging the other issue.

**Comment by Stephan202** [comment]
W.r.t. the other issue: I suspect it's due to [this code](https://github.com/openrewrite/rewrite-templating/blob/9a7f7b1402348a02f933c723bda813dbff7c9f5f/src/main/java/org/openrewrite/java/template/processor/RefasterTemplateProcessor.java#L313), which doesn't properly handle the unnamed package.

I rebased and pushed a commit to fix the build, but would like to reference an issue or PR before merging. I _may_ have time for a PR later today/this week; need to leave the train first ;)

**Comment by Stephan202** [comment]
Alright, I filed openrewrite/rewrite-templating/pull/64 and referenced it in the code.

File: error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/RefasterMethodParameterOrderTest.java
Comment (Stephan202) [reply error-prone-contrib/src/test/java/tech/picnic/errorprone/bugpatterns/RefasterMethodParameterOrderTest.java:52]: Fair point. We should add a positive case, as that verifies on which tree node the violation is reported.

**Review by Stephan202 — COMMENTED** [review]
Added one more commit.

**Comment by Stephan202** [comment]
I rebased the PR on top of #962 and reverted the package workaround. (So we should merge the upgrade PR first.)

**Comment by Stephan202** [comment]
Rebased; the PR now targets `master`. Can be merged once built.

**Pull request merged**
